impala/_thrift_api.py

```diff
@@ -346,23 +382,34 @@ def get_http_transport(host, port, http_path, timeout=None, use_ssl=False,
         except AttributeError:
             auth = base64.encodestring(user_password).decode().strip('\n')
 
-        transport.setCustomHeaders({'Authorization': 'Basic %s' % auth})
+        def get_custom_headers(cookie_headers, has_auth_cookie):
+            custom_headers = {}
+            if cookie_headers:
+                log.debug('add cookies to HTTP header')
+                custom_headers['Cookie'] = cookie_headers
+            if not has_auth_cookie:
+                custom_headers['Authorization'] = "Basic " + auth
+            return custom_headers
+
+        transport.setGetCustomHeadersFunc(get_custom_headers)
 
     elif auth_mechanism == 'GSSAPI':
         # For GSSAPI over http we need to dynamically generate custom request headers.
-        def get_auth_headers(auth_cookie):
+        def get_custom_headers(cookie_headers, has_auth_cookie):
             import kerberos
-            if auth_cookie:
-                cookie_value = auth_cookie.output(attrs=['value'], header='' ).strip()
-                return {'Cookie': cookie_value}
-            else:
+            custom_headers = {}
+            if cookie_headers:
+                log.debug('add cookies to HTTP header')
+                custom_headers['Cookie'] = cookie_headers
+            if not has_auth_cookie:
```

Comment:
Thanks for these changes. For GSSAPI requests generating the Authorization header can be slow, so I think your approach  is fine here.

(If the authentication fails because the cookie has expired on the server side, impyla can still recover by retrying the request in L312-316)